pkg/cni/types.go

```diff
@@ -0,0 +1,104 @@
+package cni
+
+import (
+	"context"
+	"github.com/containernetworking/cni/pkg/invoke"
+	"gopkg.in/k8snetworkplumbingwg/multus-cni.v3/pkg/k8sclient"
+	"gopkg.in/k8snetworkplumbingwg/multus-cni.v3/pkg/types"
+	"net/http"
+	"time"
+
+	"github.com/containernetworking/cni/pkg/types/current"
+)
+
+type cniRequestFunc func(request *PodRequest) ([]byte, error)
+
+// Explicit type for CNI commands the server handles
+type command string
+
+// PodRequest represents a request for ADD / DEL / CHECK for a Pod
+type PodRequest struct {
+	// The CNI command of the operation
+	Command command
+	// kubernetes namespace name
+	PodNamespace string
+	// kubernetes pod name
+	PodName string
+	// kubernetes pod UID
+	PodUID string
+	// kubernetes container ID
+	SandboxID string
+	// CNI container ID
+	ContainerID string
+	// kernel network namespace path
+	Netns string
+	// Interface name to be configured
+	IfName string
+	// CNI conf obtained from stdin conf
+	CNIConf *types.NetConf
+	// Timestamp when the request was started
+	timestamp time.Time
+	// ctx is a context tracking this request's lifetime
+	ctx context.Context
+	// cancel should be called to cancel this request
+	cancel context.CancelFunc
+	// kubeclient has the kubernetes API client
+	kubeclient *k8sclient.ClientInfo
+	// exec execs
+	exec invoke.Exec
+}
+
+// Request sent to the Server by the multus-shim
+type Request struct {
+	// CNI environment variables, like CNI_COMMAND and CNI_NETNS
+	Env map[string]string `json:"env,omitempty"`
+	// CNI configuration passed via stdin to the CNI plugin
+	Config []byte `json:"config,omitempty"`
+}
+
+// KubeAPIAuth contains information necessary to create a Kube API client
+type KubeAPIAuth struct {
+	// Kubeconfig is the path to a kubeconfig
+	Kubeconfig string `json:"kubeconfig,omitempty"`
+	// KubeAPIServer is the URL of a Kubernetes API server (not required if kubeconfig is given)
+	KubeAPIServer string `json:"kube-api-server,omitempty"`
+	// KubeAPIToken is a Kubernetes API token (not required if kubeconfig is given)
+	KubeAPIToken string `json:"kube-api-token,omitempty"`
+	// KubeCAData is the Base64-ed Kubernetes API CA certificate data (not required if kubeconfig is given)
+	KubeCAData string `json:"kube-ca-data,omitempty"`
+}
+
+// Server represents an HTTP server listening to a unix socket. It will handle
+// the CNI shim requests issued when a pod is added / removed.
+type Server struct {
+	http.Server
+	requestFunc       cniRequestFunc
+	rundir            string
+	useOVSExternalIDs int32
+	kubeAuth          *KubeAPIAuth
```

Comment:
`kubeAuth` may not required to keep in Server structure. Usually user gives kubeconfig (from service account), so it should be enough.